File: adventofcode/year2017/Day08.cs
using adventofcode.common;

namespace adventofcode.year2017;

public class Day08 : Solution
{
    private string[] _input;
    
    public Day08(string year, string day) : base(year, day)
    {
        _input = LoadInputAsLines();
    }

    public override string PartOne()
    {
        var cpu = new CPU(_input);
        cpu.Run();
        
        return Convert.ToString(cpu.Registers().Aggregate(0, (acc, kv) => kv.Value > acc ? kv.Value : acc));
    }

    public override string PartTwo()
    {
        var cpu = new CPU(_input);
        cpu.Run();
        
        return Convert.ToString(cpu.Largest());
    }

    private class CPU
    {
        private List<string> _instructions;
        private Dictionary<string, int> _registers;
        private int _largest;
        
        public CPU(string[] instructions)
        {
            _instructions = new List<string>(instructions);
            _registers = new Dictionary<string, int>();
            _largest = 0;
        }

        private void InitializeRegister(string register)
        {
            if (!_registers.ContainsKey(register))
            {
                _registers.Add(register, 0);
            }
        }

        private bool EvaluateCondition(string register, string condition, int value)
        {
            switch (condition)
            {
                case "<":
                    return _registers[register] < value;
                case ">":
                    return _registers[register] > value;
                case "<=":
                    return _registers[register] <= value;
                case ">=":
                    return _registers[register] >= value;
                case "==":
                    return _registers[register] == value;
                case "!=":
                    return _registers[register] != value;
                default:
                    throw new Exception($"Invalid condition : {condition}");
            }
        }

        public Dictionary<string, int> Registers()
        {
            return _registers;
        }

        public int Largest()
        {
            return _largest;
        }

        public void Run()
        {
            var index = 0;
            while (index < _instructions.Count)
            {
                var instruction = _instructions[index].Split(" ");
                var targetRegister = instruction[0];
                var operation = instruction[1];
                var targetValue = Convert.ToInt32(instruction[2]);
                var conditionRegister = instruction[4];
                var condition = instruction[5];
                var conditionValue = Convert.ToInt32(instruction[6]);
                
                InitializeRegister(targetRegister);
                InitializeRegister(conditionRegister);

                switch (operation)
                {
                    case "inc":
                        _registers[targetRegister] += EvaluateCondition(conditionRegister, condition, conditionValue) ? targetValue : 0;
                        break;
                    case "dec":
                        _registers[targetRegister] -= EvaluateCondition(conditionRegister, condition, conditionValue) ? targetValue : 0;
                        break;
                    default:
                        throw new Exception($"Invalid operation : {operation}");
                }

                _largest = _registers[targetRegister] > _largest ? _registers[targetRegister] : _largest;

                index++;
            }
        }
    }
}

File: adventofcode/year2020/day08.py
from __future__ import annotations
from adventofcode.common import Solution
from typing import List


class CPU(object):
    def __init__(self, instructions: List[str]):
        self._instructions = instructions
        self._index = 0
        self._value = 0

    @property
    def index(self) -> int:
        return self._index

    @property
    def value(self) -> int:
        return self._value

    @property
    def halted(self) -> bool:
        return not (0 <= self._index < len(self._instructions))

    def step(self):
        match self._instructions[self._index].split(" "):
            case "nop", _:
                self._index += 1
            case "acc", amount:
                self._value += int(amount)
                self._index += 1
            case "jmp", amount:
                self._index += int(amount)


class Day08(Solution):
    def __init__(self, year: str, day: str):
        super().__init__(year, day)
        self._input = self._load_input_as_lines()

    def part_one(self):
        visited = set([])
        cpu = CPU(self._input)
        value = 0
        while not cpu.halted:
            if cpu.index not in visited:
                visited.add(cpu.index)
                cpu.step()
            else:
                value = cpu.value
                break

        return value

    def part_two(self):
        value = 0
        for change_index in [i for i, instruction in enumerate(self._input) if instruction.split(" ")[0] != "acc"]:
            modified_instructions = [l for l in self._input]
            if modified_instructions[change_index].startswith("nop"):
                modified_instructions[change_index] = modified_instructions[change_index].replace("nop", "jmp")
            else:
                modified_instructions[change_index] = modified_instructions[change_index].replace("jmp", "nop")

            visited = set([])
            cpu = CPU(modified_instructions)
            while not cpu.halted:
                if cpu.index not in visited:
                    visited.add(cpu.index)
                    cpu.step()
                else:
                    break
            else:
                value = cpu.value
                break

        return value
